{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-08-02T16:07:17.380647Z",
     "start_time": "2018-08-02T16:05:47.035035Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "data/Pubmed/word_network/predict_abst_180717\n",
      "finished 0 1 mitotic nuclear division via\n",
      "interferon-tau secretion\tprepubertal g 1443402\n",
      "finished\n",
      "finished\n",
      "generate graph finished 0.0\n",
      "generate graph finished 0.24983698137\n",
      "generate graph finished 0.499673962739\n",
      "generate graph finished 0.749510944109\n",
      "generate graph finished 0.999347925479\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "\"\\nfor ci,s in enumerate(cancer_d2g):\\n    print s,len(cancer_d2g[s])\\n    for i,t in enumerate(cancer_d2g[s]):\\n        if not ImproveNet_obj.ValidPath(s,t):\\n            continue\\n        image_file = result_dir + s +'_' + t\\n        for max_ngh in range(1,10,1):            \\n            sub_graph = SubGraph(ImproveNet_obj,s,set([t]), max_path_L = max_path_L,edge_wt_thres = edge_wt_thres,max_ngh=max_ngh)\\n            gene_score[s][t],npath[s][t],edge_list,node_list = sub_graph.CalSubGraphScore(image_file = image_file, max_dup_edge_type = max_dup_edge_type,dfs_max_depth=max_path_L+1)\\n            if gene_score[s][t] > -1 and npath[s][t]>1:\\n                print s,t,gene_score[s][t],npath[s][t],t in cancer_d2g[s],max_ngh\\n                SenGene_obj.get_graph_sentence(edge_list,image_file,Net_obj)\\n                break\\n\""
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%reload_ext autoreload\n",
    "%autoreload 2\n",
    "    \n",
    "import sys\n",
    "import os\n",
    "repo_dir = '/srv/local/work/swang141/Sheng_repo/'\n",
    "nlp_tool_dir = '/srv/local/work/swang141/PatientSetAnnotation/Sheng/analysis/stanford_parser_java/stanford-parser-full-2017-06-09/'\n",
    "pubmed_dir = '/srv/local/work/swang141/PatientSetAnnotation/Sheng/data/pubmed/'\n",
    "sys.path.append(repo_dir)\n",
    "os.chdir(repo_dir)\n",
    "\n",
    "from pikachu.datasets.WordNet import WordNet\n",
    "from pikachu.datasets.ImprovedWordNet import ImprovedWordNet\n",
    "from pikachu.utils.nlp import parse_word_net\n",
    "from pikachu.models.generate_sentence.extractive_gen import ExtractGenSent\n",
    "from pikachu.datasets.SubGraph import SubGraph\n",
    "import operator\n",
    "import collections\n",
    "\n",
    "pubmed_word_net = 'data/Pubmed/word_network/predict_abst_180717'\n",
    "#pubmed_word_net = 'data/Pubmed/word_network/all_abst_180305'\n",
    "Net_obj = WordNet()\n",
    "\n",
    "Net_obj.ReadWordNet(pubmed_word_net,verbal=True,min_freq_cutoff=0)\n",
    "Net_obj.ReadWordType()\n",
    "stop_word_file = 'data/NLP_Dictionary/stopwords.txt'\n",
    "#stop_word_list = parse_word_net.GenerateStopWordList(Net_obj.word_ct,Net_obj.edge_ct,stop_word_file,Net_obj.word_type)\n",
    "stop_word_list = parse_word_net.GetStopWordList(stop_word_file)\n",
    "\n",
    "stop_word_list_manually = parse_word_net.GetStopWordList('data/NLP_Dictionary/stopwords_manually.txt')\n",
    "stop_word_list = stop_word_list.union(stop_word_list_manually)\n",
    "Net_obj.ReadWordType()\n",
    "Net_obj.ReadEdgeType(stop_word_list)\n",
    "\n",
    "candidate_gene,cancer_d2g = Net_obj.ReadNodeTypeCancerGene()\n",
    "selected_kg_l = [Net_obj.literome_g2g,Net_obj.synonym_g2g,Net_obj.hpo_d2d,Net_obj.go_f2f]\n",
    "\n",
    "ImproveNet_obj = ImprovedWordNet(Net_obj,selected_kg_l)\n",
    "\n",
    "#result_dir = '/srv/local/work/swang141/PatientSetAnnotation/Sheng/result/network_flow/cancer/'\n",
    "\n",
    "\n",
    "gene_score = collections.defaultdict(dict)\n",
    "npath = collections.defaultdict(dict)\n",
    "ImproveNet_obj.reload()\n",
    "\n",
    "'''\n",
    "for ci,s in enumerate(cancer_d2g):\n",
    "    print s,len(cancer_d2g[s])\n",
    "    for i,t in enumerate(cancer_d2g[s]):\n",
    "        if not ImproveNet_obj.ValidPath(s,t):\n",
    "            continue\n",
    "        image_file = result_dir + s +'_' + t\n",
    "        for max_ngh in range(1,10,1):            \n",
    "            sub_graph = SubGraph(ImproveNet_obj,s,set([t]), max_path_L = max_path_L,edge_wt_thres = edge_wt_thres,max_ngh=max_ngh)\n",
    "            gene_score[s][t],npath[s][t],edge_list,node_list = sub_graph.CalSubGraphScore(image_file = image_file, max_dup_edge_type = max_dup_edge_type,dfs_max_depth=max_path_L+1)\n",
    "            if gene_score[s][t] > -1 and npath[s][t]>1:\n",
    "                print s,t,gene_score[s][t],npath[s][t],t in cancer_d2g[s],max_ngh\n",
    "                SenGene_obj.get_graph_sentence(edge_list,image_file,Net_obj)\n",
    "                break\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-08-02T20:39:18.734302Z",
     "start_time": "2018-08-02T20:39:16.758986Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('transduction', -74646)\n",
      "('ion homeostasis', -63971)\n",
      "('histone acetylation', -60447)\n",
      "('adenylate kinase activity', -59504)\n",
      "('macrophage chemotaxis', -56526)\n",
      "('chemotaxis', -56432)\n",
      "('nuclear transport', -54330)\n",
      "('response to blue light', -53513)\n",
      "('adrenergic receptor binding', -50943)\n",
      "('synaptic vesicle', -50035)\n",
      "('chromosome localization', -50016)\n",
      "('erythrocyte aggregation', -47632)\n",
      "('epithelial to mesenchymal transition', -47104)\n",
      "('nucleic acid binding', -46708)\n",
      "('transcriptional attenuation', -46381)\n",
      "('regulation of dna replication', -46082)\n",
      "('spindle assembly checkpoint', -45030)\n",
      "('tolerance induction', -42068)\n",
      "('regulation of gonadotropin secretion', -42044)\n",
      "('xylanase activity', -41318)\n",
      "('regulation of synaptic plasticity', -40690)\n",
      "('ossification', -39635)\n",
      "('interleukin-4 secretion', -36810)\n",
      "('rumination', -36256)\n",
      "('acetyl-coa carboxylase activity', -35573)\n",
      "('response to corticosteroid', -33200)\n",
      "('stomatal movement', -27004)\n",
      "('response to catecholamine', -25374)\n",
      "('central vacuole', -18213)\n",
      "('diacylglycerol binding', -17871)\n",
      "('actin filament depolymerization', -16251)\n",
      "('type iv pilus biogenesis', -14561)\n",
      "('centrosome cycle', -8577)\n",
      "('skeletal system development', -6706)\n",
      "('cell leading edge', -6546)\n",
      "('response to chloramphenicol', -5920)\n",
      "('response to tunicamycin', -5742)\n",
      "('actin filament branching', -5721)\n",
      "('cst complex', -4053)\n",
      "('sublamina densa', -4031)\n",
      "('palmitoyl-coa oxidase activity', -3584)\n",
      "('mitochondrial cloud', -3286)\n",
      "('apolipoprotein binding', -2601)\n",
      "('protein trimerization', -2477)\n",
      "('amygdala development', -1646)\n",
      "('melibiose transport', -1356)\n",
      "('phosphate ion transport', -1001)\n",
      "('hemocyte migration', -608)\n",
      "('aspartoacylase activity', -553)\n",
      "('direct ossification', -502)\n",
      "('interferon-tau secretion', -365)\n",
      "('lst4-lst7 complex', -288)\n",
      "('aerobic electron transport chain', -232)\n",
      "('response to nutrient levels', -205)\n",
      "('tau protein binding', -198)\n",
      "('alpha-n-acetylglucosaminidase activity', -181)\n",
      "('maintenance of diapause', -166)\n",
      "('perianth development', -164)\n",
      "('cvt vesicle', -146)\n",
      "('regulation of cartilage development', -142)\n",
      "('regulation of chondrocyte hypertrophy', -114)\n",
      "('intermembrane lipid transfer', -85)\n",
      "('for', -67)\n",
      "('was', -66)\n",
      "('all', -62)\n",
      "('via', -60)\n",
      "('large', -56)\n",
      "('significantly higher', -54)\n",
      "('impact', -52)\n",
      "('changing', -51)\n",
      "('great', -51)\n",
      "('examine', -51)\n",
      "('set', -50)\n",
      "('find', -50)\n",
      "('elucidate', -50)\n",
      "('science', -50)\n",
      "('inc', -50)\n",
      "('significantly lower', -50)\n",
      "('significant differences', -50)\n",
      "('chain', -49)\n",
      "('cyclin-dependent protein kinase activity', -49)\n",
      "('stimulate', -49)\n",
      "('significantly reduced', -49)\n",
      "('fact', -49)\n",
      "('cancer', -48)\n",
      "('dependence', -48)\n",
      "('occurs in', -48)\n",
      "('named', -48)\n",
      "('susceptible', -48)\n",
      "('gain', -48)\n",
      "('kinds', -48)\n",
      "('defects', -48)\n",
      "('minor', -48)\n",
      "('correlates', -47)\n",
      "('epithelial cells', -47)\n",
      "('block', -47)\n",
      "('outside', -47)\n",
      "('atp', -47)\n",
      "('blocks', -47)\n",
      "('disruption', -47)\n",
      "False\n",
      "('for', 8)\n",
      "('was', 7)\n",
      "('tau aggregates', 4)\n",
      "('tau isoforms', 4)\n",
      "('4r tau', 3)\n",
      "('microtubules', 3)\n",
      "('all', 2)\n",
      "('tau species', 2)\n",
      "('mobility shift', 2)\n",
      "('phosphorylated tau', 2)\n",
      "('gel electrophoresis', 2)\n",
      "('microtubule binding', 2)\n",
      "('3r tau', 2)\n",
      "('pairs', 2)\n",
      "('tau aggregation', 2)\n",
      "('ftdp-17', 2)\n",
      "('paired helical filaments', 2)\n",
      "('microtubule-associated protein tau', 2)\n",
      "('significantly higher', 2)\n",
      "('biochemical analysis', 2)\n",
      "('n-terminal', 2)\n",
      "('microtubule binding domains', 2)\n",
      "('emerging', 1)\n",
      "('acts', 1)\n",
      "('microtubule cytoskeleton', 1)\n",
      "('lewis', 1)\n",
      "('extremely low', 1)\n",
      "('stabilizes microtubules', 1)\n",
      "('pka', 1)\n",
      "('neuronal morphogenesis', 1)\n",
      "('coli', 1)\n",
      "('twofold greater', 1)\n",
      "('find', 1)\n",
      "('disorders', 1)\n",
      "('susceptible', 1)\n",
      "('e10', 1)\n",
      "('e12', 1)\n",
      "('tau fragments', 1)\n",
      "('ivig products', 1)\n",
      "('tryptophan', 1)\n",
      "('fivefold higher', 1)\n",
      "('previous reports', 1)\n",
      "('scaffold protein', 1)\n",
      "('disulfide', 1)\n",
      "('tg', 1)\n",
      "('equilibrium', 1)\n",
      "('steady-state', 1)\n",
      "('presently unclear', 1)\n",
      "('extracellular tau', 1)\n",
      "('tau fragment', 1)\n",
      "('tau oligomers', 1)\n",
      "('differentially phosphorylated', 1)\n",
      "('nft pathology', 1)\n",
      "('phosphorylation site', 1)\n",
      "('widely held', 1)\n",
      "('cytoplasmic protein', 1)\n",
      "('steady state', 1)\n",
      "('estramustine phosphate', 1)\n",
      "('total tau protein', 1)\n",
      "('kinase-dead', 1)\n",
      "('amino acid analysis', 1)\n",
      "('microtubule assembly', 1)\n",
      "(\"alzheimer's disease brains\", 1)\n",
      "('straight filaments', 1)\n",
      "('14-3-3zeta', 1)\n",
      "('micromolar concentrations', 1)\n",
      "('strong binding', 1)\n",
      "('greatly simplified', 1)\n",
      "('polymers', 1)\n",
      "('l is', 1)\n",
      "('fast protein liquid chromatography', 1)\n",
      "('toxic oligomers', 1)\n",
      "('ser', 1)\n",
      "('glycogen-synthase-kinase-3beta', 1)\n",
      "('kda', 1)\n",
      "('parallel arrays', 1)\n",
      "('antibody levels', 1)\n",
      "('fibroblast cells', 1)\n",
      "('supranuclear palsy', 1)\n",
      "('sodium dodecyl', 1)\n",
      "('ammonium sulfate precipitation', 1)\n",
      "('ion exchange', 1)\n",
      "('slightly decreased', 1)\n",
      "('tau transgenic mice', 1)\n",
      "('large', 1)\n",
      "('phosphorylation state', 1)\n",
      "('protein kinases', 1)\n",
      "('tau kinase', 1)\n",
      "('sporadic ftld', 1)\n",
      "('microtubule disrupting drugs', 1)\n",
      "('sulfhydryl groups', 1)\n",
      "('g2019s mutant', 1)\n",
      "('microdialysis technique', 1)\n",
      "('human tau protein', 1)\n",
      "('elevated levels', 1)\n",
      "('tau protein kinase ii', 1)\n",
      "('microtubule-binding domain', 1)\n",
      "('individual species', 1)\n",
      "('length distributions', 1)\n",
      "('assembly', 1)\n",
      "('antibody concentrations', 1)\n",
      "('leucine-rich repeat kinase 2', 1)\n",
      "('heterogeneous', 1)\n",
      "('bovine brain', 1)\n",
      "('specific activity', 1)\n",
      "('tau protein kinase', 1)\n",
      "('knocking down', 1)\n",
      "('control subjects', 1)\n",
      "('human neuroblastoma', 1)\n",
      "('seeds', 1)\n",
      "('atp', 1)\n",
      "('act', 1)\n",
      "('microtubule polymerization', 1)\n",
      "('gammagard', 1)\n",
      "('isf', 1)\n",
      "('hyperphosphorylation of tau', 1)\n",
      "('named', 1)\n",
      "('nacl', 1)\n",
      "('extracellular fluids', 1)\n",
      "('tightly packed', 1)\n",
      "('basic ph', 1)\n",
      "('stabilize microtubules', 1)\n",
      "('brain homogenates', 1)\n",
      "('highly reactive', 1)\n",
      "('or 2', 1)\n",
      "('filament growth', 1)\n",
      "('mrna levels', 1)\n",
      "('58-amino-acid', 1)\n",
      "('aggregated tau', 1)\n",
      "('isoform composition', 1)\n",
      "('biol', 1)\n",
      "('strongly reactive', 1)\n",
      "('repeat sequences', 1)\n",
      "('human tau isoforms', 1)\n",
      "('tubulin assembly', 1)\n",
      "('neuroblastoma cells', 1)\n",
      "('c-terminus', 1)\n",
      "('purification method', 1)\n",
      "('size exclusion', 1)\n",
      "('amino terminal', 1)\n",
      "('tubulin polymerization', 1)\n",
      "('protein it', 1)\n",
      "('progressive supranuclear', 1)\n",
      "('paired helical filament tau', 1)\n",
      "('anatomical areas', 1)\n",
      "('phosphoproteins', 1)\n",
      "('amorphous aggregates', 1)\n",
      "('structurally stable', 1)\n",
      "('binding affinity', 1)\n",
      "('chromatographic techniques', 1)\n",
      "('cysteine', 1)\n",
      "('insoluble tau', 1)\n",
      "('phosphoserine', 1)\n",
      "('purified tubulin', 1)\n",
      "('soluble tau', 1)\n",
      "('58-amino acid', 1)\n",
      "('hydroxyapatite chromatography', 1)\n",
      "('cam kinase ii', 1)\n",
      "('m17', 1)\n",
      "('e2 e3', 1)\n",
      "('hippocampal ca1 pyramidal neurons', 1)\n",
      "('csf tau', 1)\n",
      "('experimental treatments', 1)\n",
      "('phosphorylate', 1)\n",
      "('parkinsonism linked', 1)\n",
      "('giving rise', 1)\n",
      "('epitopes', 1)\n",
      "('highly phosphorylated tau', 1)\n",
      "('microtubule bundles', 1)\n",
      "('significant differences', 1)\n",
      "('human tau', 1)\n",
      "('closely related', 1)\n",
      "('neuronal metabolism', 1)\n",
      "('phfs', 1)\n",
      "('protein levels', 1)\n",
      "('double stranded', 1)\n",
      "('correlates', 1)\n",
      "('hyper-phosphorylated tau', 1)\n",
      "('prerequisite', 1)\n",
      "('bovine brain microtubules', 1)\n",
      "('specific binding', 1)\n",
      "('phosphorylates tau', 1)\n",
      "('brain interstitial fluid', 1)\n",
      "('lrrk2 kinase activity', 1)\n",
      "('intermolecular disulfide', 1)\n",
      "('changing', 1)\n",
      "('normal tau', 1)\n",
      "('tpki', 1)\n",
      "('microtubule formation', 1)\n",
      "('recombinant human tau', 1)\n",
      "('acids', 1)\n",
      "('biochemistry', 1)\n",
      "('microtubule binding domain', 1)\n",
      "('ck-1', 1)\n",
      "('lobar degeneration', 1)\n",
      "('mechanistic insights', 1)\n",
      "('familial frontotemporal dementia', 1)\n",
      "('paired', 1)\n"
     ]
    }
   ],
   "source": [
    "term = Net_obj.edge_ct.keys()\n",
    "term_ct = {}\n",
    "for i in term:\n",
    "    term_ct[i] = len(Net_obj.edge_ct[i])*-1\n",
    "term_ct_s = sorted(term_ct.items(), key=lambda kv: kv[1])\n",
    "#print len(term_ct_s),len(term_ct)\n",
    "for i in range(100):\n",
    "    print term_ct_s[i]\n",
    "    \n",
    "term = 'tau protein binding'\n",
    "\n",
    "print 'lrrk2' in Net_obj.edge_ct[term]\n",
    "go_ct_s = sorted(Net_obj.edge_ct[term].items(), key=lambda kv: kv[1], reverse=True)\n",
    "for i in range(len(go_ct_s)):\n",
    "    print go_ct_s[i]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'ImproveNet_obj' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-2-fe0203c8aeed>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     10\u001b[0m \u001b[0mmax_dup_edge_type\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m2\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 12\u001b[0;31m \u001b[0msub_graph\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mSubGraph\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mImproveNet_obj\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0ms\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mt\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmax_path_L\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmax_path_L\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0medge_wt_thres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0medge_wt_thres\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mmax_ngh\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mmax_ngh\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     13\u001b[0m \u001b[0mgene_score\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mnpath\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0medge_list\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mnode_list\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msub_graph\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mCalSubGraphScore\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mimage_file\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mimage_file\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmax_dup_edge_type\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmax_dup_edge_type\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mdfs_max_depth\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mmax_path_L\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     14\u001b[0m \u001b[0mSenGene_obj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_graph_sentence\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0medge_list\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mimage_file\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mNet_obj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'ImproveNet_obj' is not defined"
     ]
    }
   ],
   "source": [
    "result_dir = '/srv/local/work/swang141/Sheng_repo/result/network_flow/cancer/'\n",
    "SenGene_obj = ExtractGenSent(repo_dir,nlp_tool_dir,pubmed_dir)\n",
    "t = ['bptf','tp53']\n",
    "s = 'liver hepatocellular carcinoma'\n",
    "image_file = result_dir + s +'_' + t[0]\n",
    "\n",
    "max_ngh = 200\n",
    "max_path_L = 4\n",
    "edge_wt_thres = 1\n",
    "max_dup_edge_type = 2\n",
    "\n",
    "sub_graph = SubGraph(ImproveNet_obj,s,set(t), max_path_L = max_path_L,edge_wt_thres = edge_wt_thres,max_ngh=max_ngh)\n",
    "gene_score,npath,edge_list,node_list = sub_graph.CalSubGraphScore(image_file = image_file, max_dup_edge_type = max_dup_edge_type,dfs_max_depth=max_path_L+1)\n",
    "SenGene_obj.get_graph_sentence(edge_list,image_file,Net_obj)\n",
    "print SenGene_obj.edge2sent"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "%reload_ext autoreload\n",
    "%autoreload 2\n",
    "    \n",
    "import sys\n",
    "import os\n",
    "\n",
    "#repo_dir = '/srv/local/work/swang141/Sheng_repo/'\n",
    "repo_dir = '/data/cellardata/users/majianzhu/wangsheng/Sheng_repo/Sheng_repo/'\n",
    "\n",
    "sys.path.append(repo_dir)\n",
    "os.chdir(repo_dir)\n",
    "from pikachu.models.generate_sentence.preprocess_parser_tree import preprocess_parser_tree\n",
    "from pikachu.models.generate_sentence.preprocess_word_sent import get_sentence_edge\n",
    "sent_l = get_sentence_edge(\"dysplastic_nevus_syndrome\", 'bap1')\n",
    "\n",
    "parser_obj = preprocess_parser_tree()\n",
    "vocab_score,path_l,stem_word_list_l = parser_obj.get_parser_tree(\"dysplastic_nevus_syndrome\", 'bap1',sent_l)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "x = []\n",
    "for d in cancer_d2g:\n",
    "    x.append(len(cancer_d2g[d]))\n",
    "\n",
    "print x\n",
    "print np.mean(np.array(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "ct = 0\n",
    "for a in ImproveNet_obj.net:\n",
    "    ct += len(ImproveNet_obj.net[a])\n",
    "print ct"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "print 'this is a \" book'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "a='sb'\n",
    "\n",
    "print a+'\"'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import scipy.stats as stats\n",
    "tab = [[100,1],[2,100]]\n",
    "odds,pv = stats.fisher_exact(tab,alternative='greater')\n",
    "print odds,pv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import torchtext"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pos = data.TabularDataset(\n",
    "   path='data/pos/pos_wsj_train.tsv', format='tsv',\n",
    "fields=[('text', data.Field()),\n",
    "            ('labels', data.Field())])"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "print len(ImproveNet_obj.net)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-08-02T18:30:29.577784Z",
     "start_time": "2018-08-02T18:30:12.663202Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "set(['unknown', 'mesh_function', 'Han_function', 'mesh_drug', 'mesh_entity'])\n",
      "hpo_disease Number of words: 340 cutoff 10\n",
      "('binding sites', 7918)\n",
      "('chromosomes', 5609)\n",
      "('molecular weight', 4412)\n",
      "('hydrolysis', 3663)\n",
      "('up-regulation', 3336)\n",
      "('exons', 2915)\n",
      "('radiation', 2423)\n",
      "('mutagenesis', 2212)\n",
      "('filtration', 2058)\n",
      "('operon', 1998)\n",
      "('transfection', 1939)\n",
      "('acetylation', 1863)\n",
      "('hematocrit', 1522)\n",
      "('plasmids', 1509)\n",
      "('virulence', 1509)\n",
      "('half-life', 1407)\n",
      "('introns', 1379)\n",
      "('permeability', 1150)\n",
      "('substrate specificity', 1117)\n",
      "('pulse', 1108)\n",
      "('consensus sequence', 1092)\n",
      "('sequence homology', 1048)\n",
      "('open reading frames', 1018)\n",
      "('chromatids', 984)\n",
      "('homologous recombination', 941)\n",
      "('tata box', 936)\n",
      "('alternative splicing', 810)\n",
      "('genomic library', 796)\n",
      "('catalytic domain', 748)\n",
      "('transcription initiation site', 734)\n",
      "(\"5' flanking region\", 675)\n",
      "('prometaphase', 618)\n",
      "('replication origin', 606)\n",
      "('follicular phase', 598)\n",
      "('luteal phase', 584)\n",
      "('cell cycle checkpoints', 552)\n",
      "('adsorption', 550)\n",
      "('radioactivity', 539)\n",
      "('catalysis', 529)\n",
      "('color', 512)\n",
      "('tissue distribution', 489)\n",
      "('electron transport', 488)\n",
      "('compliance', 469)\n",
      "('germination', 463)\n",
      "('expressed sequence tags', 435)\n",
      "('x-rays', 417)\n",
      "('erythrocyte deformability', 412)\n",
      "('telomere', 372)\n",
      "('dna breaks', 352)\n",
      "('response elements', 349)\n",
      "('polyadenylation', 348)\n",
      "('sumoylation', 348)\n",
      "('multigene family', 341)\n",
      "('protons', 334)\n",
      "('salinity', 317)\n",
      "('secretory pathway', 315)\n",
      "('nuclear localization signals', 306)\n",
      "('replicon', 297)\n",
      "('sound', 297)\n",
      "('conserved sequence', 287)\n",
      "('gene order', 274)\n",
      "('x-ray diffraction', 269)\n",
      "('solubility', 267)\n",
      "('hemorheology', 265)\n",
      "('excitatory postsynaptic potentials', 257)\n",
      "('fluorescence resonance energy transfer', 253)\n",
      "('antibody formation', 249)\n",
      "('polyploidy', 245)\n",
      "('protein stability', 238)\n",
      "('cell count', 238)\n",
      "('histocompatibility', 237)\n",
      "('protein domains', 234)\n",
      "('synteny', 225)\n",
      "('pseudogenes', 223)\n",
      "('haplotypes', 221)\n",
      "('cosmids', 214)\n",
      "('genetic markers', 206)\n",
      "('ion exchange', 206)\n",
      "('central venous pressure', 205)\n",
      "('proto-oncogenes', 197)\n",
      "('blood cell count', 197)\n",
      "('hydroxylation', 195)\n",
      "('gene dosage', 194)\n",
      "('zinc fingers', 189)\n",
      "('gametogenesis', 187)\n",
      "('alkylation', 183)\n",
      "('genomic imprinting', 183)\n",
      "('photons', 182)\n",
      "('gene duplication', 182)\n",
      "('biotransformation', 180)\n",
      "('polytene chromosomes', 165)\n",
      "('dna cleavage', 163)\n",
      "('gene expression regulation', 162)\n",
      "('bacterial load', 161)\n",
      "('cell fusion', 157)\n",
      "('structure-activity relationship', 154)\n",
      "('deamination', 152)\n",
      "('catabolite repression', 149)\n",
      "('chimerism', 147)\n",
      "('energy transfer', 147)\n"
     ]
    }
   ],
   "source": [
    "fin = open('data/NLP_Dictionary/stopwords.txt')\n",
    "t = set()\n",
    "gct = {}\n",
    "for line in fin:\n",
    "    w = line.strip().split('\\t')\n",
    "    t.add(w[1])\n",
    "    if w[1] == 'mesh_function':\n",
    "        gct[w[0]] = int(w[3])\n",
    "        #print line,\n",
    "fin.close()\n",
    "print t\n",
    "\n",
    "ignore_node_type = ['cancer_gene','cancer_disease','literome_gene', 'barbasi_tissue','CCLE_gene', 'mesh_disease', 'mesh_tissue','hpo_disease','GO_function','ncbi_gene']\n",
    "\n",
    "stop_word_list = parse_word_net.GenerateStopWordList(Net_obj.word_ct,\n",
    "                                                     Net_obj.edge_ct,stop_word_file,Net_obj.word_type,ignore_node_type=ignore_node_type)\n",
    "\n",
    "#ignore_node_type = ['','literome_gene', 'barbasi_tissue','CCLE_gene', 'mesh_disease', 'mesh_tissue','hpo_disease','GO_function','ncbi_gene']go_ct_s = sorted(gct.items(), key=lambda kv: kv[1], reverse=True)\n",
    "for i in range(100):\n",
    "    print go_ct_s[i]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-08-02T18:26:21.218241Z",
     "start_time": "2018-08-02T18:26:21.191945Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'data/NLP_Dictionary/stopwords.txt'"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stop_word_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
